035-ca-abstraction.md

```diff
@@ -0,0 +1,315 @@
+
+```
+  NEW // a certificate is needed for the end entity, but doesn't exist
+   |
+   v
+  REQUESTED // a certificate has been requested
+   |
+   v
+  PENDING // a certificate has been issued, but the root CA is not trusted yet
+   |
+   v
+  IN_USE // the certificate is in use
+   |
+   v
+  NOT_NEEDED // a certificate for the end entity is not needed any more
+```
+
+Then the "CA certificate trust state machine":
+
+```
+  UNTRUSTED // not yet trusted everywhere it needs to be 
+   |
+   v
+  TRUSTED_NEW // trusted everywhere it needs to be, but no EE certificates yet issued
+   |
+   v
+  TRUSTED_USED // trusted everywhere, EE certificates issued
+   |
+   v
+  PHASE_OUT // trusted everywhere, but being phased out
+   |
+   v
+  ZOMBIE // still trusted by some component, but not relied on by any party.
+```
+
+Now let's describe the processes.
+
+1. The issuing operator needs to figure out which EE certificates are missing (or in their renewal period, or otherwise need replacing, e.g. via annotation). This need for a certificate might exist independently of the end entity. For example, if we're going to scale up the Kafka cluster then we need a certificate for the new broker before the new broker exists. **The issuing operator puts the end entity in the `NEW` state.**
+2. (**See step 1 in the diagram.**) The issuing operator requests a new certificate. **The issuing operator puts the end entity in the `REQUESTED` state.**
+3. (**See step 2 in the diagram.**) The issuing operator cannot assume that issuance will happen quickly (i.e. that `pollForRequestedCertificate()` returns a non-empty `Optional` immediately), so it needs to be able persist the token obtained from `requestCertificate()` in a `Secret`, _and possibly complete the current reconciliation_ (i.e. issuance can be very asynchronous). 
+(An example of how this could happen is where a `CertificateIssuer` implementation uses cert-manager; `requestCertificate()` creates a cert-manager `CertificateRequest` CR but while the cert manager operator is not running no certificate will be issued).
+
+4. Once a certificate is issued, the issuing operator needs check whether the CA certificate is trusted. 
+   I.e. it needs to know what trust states all the existing CA certificates are in. 
+   If it is not trusted **the issuing operator puts the end entity in the `PENDING` state.** and **puts the CA certificate in the `UNTRUSTED` state**. (**See step 3 in the diagram.**)
+
+> From the above it is clear that there is a `Secret` which stores tokens and issued certificates and their current state, and that only the issuing operator needs access to it.
+> It's also clear that there is a `Secret` which stores CA certificates, and their current state, and which needs to be accessed by the issuing operator and the cluster operator.
+
+#### Process for ensuring CA trust
+
+When an EE certificate is issued via a `CertificateIssuer` it might have been signed by a CA that is not yet trusted.
```

Comment:
We could certainly add an allowlist of trusted CA certificates, and not trust CA certificates not in that list. We'd need a `NOT_ALLOWED` state in the CA certificate trust state machine. Then the user would need to take some action, since requesting another certificate from that issuer is likely to result in the same unacceptable CA certificate being used. Would that be sufficent @scholzj?

Reply:
Does the person configuring the `Kafka` CR _necessarily_ know the root CA certificate which their configured certificate issuer is going to use? If they do then they can configure the `Kafka` CR to trust EE certificates signed with that root CA certificate. But if they don't then they can't configure the `Kafka` CR at all. 

Even when they do know, it creates extra problems: What should happen when that root CA certificate expires? The operator then is stuck. So we need extra management code to add warning conditions when that's going to happen etc. I'm sure we can do that, but I'm not sure we want the complication in the initial implementation.

Reply:
It is not necessarily about the person knowing. The provider implementation should know. As explain in the main comment => you already have a chunk of this information in the CA configuration. In case of the internal issuer, it would be just the Kubernetes Secret as today. In case of the CertManager, it is the Issuer what defines what CA is used. So the provider implementation knows how to get it.

Imagine the difference like this:
* If you take the CAs from the issued certificate, as a bad actor I need just to subvert one issued certificate to inject Strimzi with my own CA and gain access. The only way you would notice is by checking the CAs trusted by Strimzi Kafka cluster.
* If you take the CA from the provider implementation instead of from the issued certificates, the bad actor would need to modify the Issuer to get Strimzi to trust some new CA. That has IMHO better chance of being noticed because the issuer might be used by other apps as well etc.

Reply:
Indeed. But on the other hand, if I'm building a cluster where I want to have a complete trust chain across all components, I want the ability to supply my own intermediate ca with a key via cluster issuer.

Your question goes both ways: can I trust the CA that comes with Strimzi?